### PR TITLE
Feature/285 lazy loading collection

### LINKED ISSUE #285: LazyLoading bug for SPARQL based Attributes
If the @Sparql annotation is used to define an attribute that can be lazily resolved with a SPARQL query, JOPA throws an exception:

```java
@Sparql(
    query = "SELECT ?capacity WHERE { ?x a <http://uob.iodt.ibm.com/univ-bench-dl.owl#Course> . ?x <http://uob.iodt.ibm.com/univ-bench-dl.owl#capacity> ?capacity }",
    fetchType = FetchType.LAZY
)
private Collection<Integer> allCapacity;
```

```
Exception in thread "main" java.lang.ClassCastException: class cz.cvut.kbss.jopa.model.metamodel.PluralQueryAttributeImpl cannot be cast to class cz.cvut.kbss.jopa.model.metamodel.PluralAttribute (cz.cvut.kbss.jopa.model.metamodel.PluralQueryAttributeImpl and cz.cvut.kbss.jopa.model.metamodel.PluralAttribute are in unnamed module of loader 'app')
	at cz.cvut.kbss.jopa.proxy.lazy.LazyLoadingProxyFactory.createProxy(LazyLoadingProxyFactory.java:62)
	at cz.cvut.kbss.jopa.sessions.CloneBuilder.populateAttributes(CloneBuilder.java:168)
	at cz.cvut.kbss.jopa.sessions.CloneBuilder.buildCloneImpl(CloneBuilder.java:136)
	at cz.cvut.kbss.jopa.sessions.CloneBuilder.buildClone(CloneBuilder.java:83)
	at cz.cvut.kbss.jopa.sessions.AbstractUnitOfWork.registerExistingObject(AbstractUnitOfWork.java:457)
	at cz.cvut.kbss.jopa.sessions.AbstractUnitOfWork.readObjectInternal(AbstractUnitOfWork.java:288)
	at cz.cvut.kbss.jopa.sessions.AbstractUnitOfWork.readObject(AbstractUnitOfWork.java:272)
	at cz.cvut.kbss.jopa.model.EntityManagerImpl.find(EntityManagerImpl.java:296)
	at cz.cvut.kbss.jopa.model.EntityManagerImpl.find(EntityManagerImpl.java:281)
	at cz.cvut.kbss.jopa.example01.ManualRunner.execute(ManualRunner.java:73)
	at cz.cvut.kbss.jopa.example01.ManualRunner.run(ManualRunner.java:57)
	at cz.cvut.kbss.jopa.example01.Example.main(Example.java:45)
```

### PR DESCRIPTION
**Pull request by luxbe**
This fixes #285 by properly resolving the Collection type for not only `PluralAttribute`s but also `PluralQueryAttribute`s

### CODE REVIEW

File: jopa-impl/src/main/java/cz/cvut/kbss/jopa/proxy/lazy/LazyLoadingProxyFactory.java [hunk 1]
```diff
@@ -59,8 +61,18 @@ public <T> Object createProxy(T entity, FieldSpecification<? super T, ?> fieldSp
         } else if (Map.class.isAssignableFrom(type)) {
             return new LazyLoadingMapProxy<>(entity, (FieldSpecification) fieldSpec, uow);
         } else if (Collection.class.isAssignableFrom(type)) {
-            final PluralAttribute<? super T, ?, ?> pa = (PluralAttribute<? super T, ?, ?>) fieldSpec;
-            return switch (pa.getCollectionType()) {
+            CollectionType collectionType;
```
Comment (ledsoft) [inline jopa-impl/src/main/java/cz/cvut/kbss/jopa/proxy/lazy/LazyLoadingProxyFactory.java:64]: How does this change the behavior? The only difference I see is that the `pa.getCollectionType()` is assigned to a variable after `fieldSpec` being cast to a more specific type. This is IMHO functionally equivalent to the current code.

**Review by ledsoft — COMMENTED** [review]
I don't see any functional changes to the code.

Furthermore, I would like a test to evaluate the changes.

File: jopa-impl/src/main/java/cz/cvut/kbss/jopa/proxy/lazy/LazyLoadingProxyFactory.java (hunk 1 above)
Comment (luxbe) [reply jopa-impl/src/main/java/cz/cvut/kbss/jopa/proxy/lazy/LazyLoadingProxyFactory.java:64]: Before, `fieldSpec` was always cast to the `PluralAttribute` type. However, it was also possible that `fieldSpec` was of type `PluralQueryAttribute` which is - despite the similiar name - not a subclass of PluralAttribute which would result in a `ClassCastException` being thrown as mentioned in the original issue

Comment (ledsoft) [reply jopa-impl/src/main/java/cz/cvut/kbss/jopa/proxy/lazy/LazyLoadingProxyFactory.java:64]: Yes, you are correct. I misunderstood the issue.

File: jopa-impl/src/test/java/cz/cvut/kbss/jopa/environment/OWLClassV.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        return OWLClassV.class.getDeclaredField("collectionStringAtt");
+    }
+
+    public static Field getSetQueryStringAttField() throws NoSuchFieldException {
+        return OWLClassV.class.getDeclaredField("setQueryStringAtt");
+    }
+
+    public static Field getListQueryStringAttField() throws NoSuchFieldException {
+        return OWLClassV.class.getDeclaredField("listQueryStringAtt");
+    }
```
Comment (ledsoft) [inline jopa-impl/src/test/java/cz/cvut/kbss/jopa/environment/OWLClassV.java:89]: Please move the static fields getters to the end of the class.

**Review by ledsoft — COMMENTED** [review]
Just one minor adjustment.

**Pull request merged**